drivers/i2c/i2c_xilinx.c

```diff
@@ -0,0 +1,641 @@
+
+	uint32_t events = i2c_xilinx_wait_interrupt(
+		config, data, finish_bit | ISR_TX_ERR_TARGET_COMP | ISR_ARB_LOST);
+	if (events != finish_bit) {
+		if (!events) {
+			return -ETIMEDOUT;
+		}
+		if (events & ISR_ARB_LOST) {
+			LOG_ERR("Arbitration lost on TX");
+			return -EAGAIN;
+		}
+		LOG_ERR("TX received NAK");
+		return -ENXIO;
+	}
+	return 0;
+}
+
+static int i2c_xilinx_wait_not_busy(const struct i2c_xilinx_config *config,
+				    struct i2c_xilinx_data *data)
+{
+	if (sys_read32(config->base + REG_SR) & SR_BB) {
+		uint32_t events = i2c_xilinx_wait_interrupt(config, data, ISR_BUS_NOT_BUSY);
+
+		if (events != ISR_BUS_NOT_BUSY) {
+			LOG_ERR("Bus stuck busy");
+			i2c_xilinx_reinit(config);
+			return -EBUSY;
+		}
+	}
+	return 0;
+}
+
+static int i2c_xilinx_write(const struct i2c_xilinx_config *config, struct i2c_xilinx_data *data,
+			    const struct i2c_msg *msg, uint16_t addr)
+{
+	const uint8_t *write_ptr = msg->buf;
+	uint32_t bytes_left = msg->len;
+	uint32_t cr = CR_EN | CR_TX;
+	uint32_t fifo_space = FIFO_SIZE - 1; /* account for address being written */
+
+	if (msg->flags & I2C_MSG_RESTART) {
+		cr |= CR_MSMS | CR_RSTA;
+	}
+
+	i2c_xilinx_clear_interrupt(config, data, ISR_TX_ERR_TARGET_COMP | ISR_ARB_LOST);
+
+	sys_write32(cr, config->base + REG_CR);
+	sys_write32((addr << 1) | TX_FIFO_START, config->base + REG_TX_FIFO);
+
+	/* TX FIFO empty detection is somewhat fragile because the status register
+	 * TX_FIFO_EMPTY bit can be set prior to the transaction actually being
+	 * complete, so we have to rely on the TX empty interrupt.
+	 * However, delays in writing data to the TX FIFO could cause it
+	 * to run empty in the middle of the process, causing us to get a spurious
+	 * completion detection from the interrupt. Therefore we disable interrupts
+	 * while the TX FIFO is being filled up to try to avoid this.
+	 */
+
+	while (bytes_left) {
+		uint32_t bytes_to_send = bytes_left;
+		const unsigned int key = irq_lock();
+		int ret;
+
+		if (bytes_to_send > fifo_space) {
+			bytes_to_send = fifo_space;
+		}
+		while (bytes_to_send) {
+			uint32_t write_word = *write_ptr++;
+
+			if (bytes_left == 1 && (msg->flags & I2C_MSG_STOP)) {
+				write_word |= TX_FIFO_STOP;
+			}
+			sys_write32(write_word, config->base + REG_TX_FIFO);
+			bytes_to_send--;
+			bytes_left--;
+		}
+		i2c_xilinx_clear_interrupt(config, data, ISR_TX_FIFO_EMPTY | ISR_BUS_NOT_BUSY);
+		irq_unlock(key);
+
+		ret = i2c_xilinx_wait_tx_done(config, data,
+					      !bytes_left && (msg->flags & I2C_MSG_STOP));
+		if (ret) {
+			return ret;
+		}
+		fifo_space = FIFO_SIZE;
+	}
+	return 0;
+}
+
+static int i2c_xilinx_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			       uint16_t addr)
+{
+	const struct i2c_xilinx_config *config = dev->config;
+	struct i2c_xilinx_data *data = dev->data;
+	int ret;
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	ret = i2c_xilinx_wait_not_busy(config, data);
```

Comment:
This is going to wait for an interrupt or 100ms, that seems like a pretty long time if there's no interrupt here? Unless I've mis-read. Aren't you already guaranteed that the IP block is idle at this point?

Reply:
The IP block itself should be idle (unless perhaps there was a target-mode transfer in progress) but it's possible another controller using the bus when the transfer is started, so this waits for that to be complete. 100ms might be excessive, I basically picked that as an arbitrary value for "if it takes this long something is almost certainly broken" - open to another suggestion..